### ISSUE TITLE
Attribute based ABI opt-in

### ISSUE THREAD

**Issue by SidneyCogdill** [SebastienBini/cpp-relocation-proposal#39]
P2785R4 adds new rules when `reloc` should not call relocate ctor to solve the ABI issue. However, this significantly complicates the design and hurts teachability, causing it to suffer from the same "std::move doesn't move" issue. Furthermore, the early destruction use-case is no longer reliably usable:

```cpp
auto read_and_update() {
    std::lock_guard l{m_mtx};
    auto num = m_res.load();
    reloc l; // Early dropping the lock guard without requiring std::unique_lock.
    very_slow_update(num); // Assuming this a blocking call that takes a long time...
}
```

In the above case, if it's decided that `reloc` can be silently deferred to later stage, it can introduce a silent performance regression without the users knowing, which in turn causes the ecosystem to simply not trust the effectiveness of discard semantic, resulting in a mechanism that looks good on paper but gains none of real adoption.

I suggest that `reloc` **mandates** the operand to be an owned parameter, or is ill-formed (emits a compile time error). And use a new attribute to explicitly opt into the new ABI behavior:

```cpp
void f2(std::unique_ptr<int> p);

auto f1(std::unique_ptr<int> [[owned]] p) {
    f2(reloc p);
}
```

Which forces `p` to be an owned parameter and callee-destroyed. This makes the ABI breakage strictly opt-in without hurting the explicitness of `reloc` operator.

**Comment by SebastienBini** [comment]
Hello, I think you are overstating the issue. P2785R4 states that:
- if the variable is not a function parameter, then the relocation is not silently discarded. Likewise, discarded reloc expressions are guaranteed to call the destructor
- if the variable is a function parameter passed by value, and its type provides a relocation constructor and no move constructor, then we have the guarantees as above

On discarded reloc expressions for value parameters:
- if the type provides a move constructor, then implementations are allowed to postpone the destruction of the object until the function exits (so to preserve the ABI). However, those expressions are not a no-op: in that case, a temporary object is move-constructed and destroyed. This will empty the source object of its resources and trigger the desired destructor effect.
- if the type provides a copy constructor and no relocation constructor, then such expressions are no-op and compilers are encouraged to emit a warning.

In other words you get:
- direct destructor call (relocation ctor and no move ctor)
- move-constructed temporary and destruction (move ctor)
- a warning which will likely be caught by Werror

Your suggestion with the `[[owned]]` attribute was already considered and was not favored. The main problem is that now `[[owned]]` is part of the function signature, and needs to be carried over to function pointers.
`void (*f)(std::unique_ptr<int> [[owned]]) = f1;`
It's a lot of boilerplate that will complicate the use of the feature and will be useless for some users as their build system already uses callee-destroy ABI.
